### PR TITLE
fix metricscache location bug

### PR DESCRIPTION
**Pull request by huijunw**
the metricscache location is an ephemeral node in zk. 
when the zk session re-connection happens, the ephemeral node may need registration again.

### CODE REVIEW

File: heron/statemgrs/src/java/com/twitter/heron/statemgr/zookeeper/curator/CuratorStateManager.java [hunk 1]
```diff
@@ -337,6 +339,16 @@ protected Lock getLock(String path) {
   public ListenableFuture<Boolean> setMetricsCacheLocation(
       TopologyMaster.MetricsCacheLocation location,
       String topologyName) {
+    client.getConnectionStateListenable().addListener(new ConnectionStateListener() {
+      @Override
+      public void stateChanged(CuratorFramework arg0, ConnectionState arg1) {
+        if (arg1 == ConnectionState.RECONNECTED || arg1 == ConnectionState.CONNECTED) {
+          LOG.info("zk session state changed to " + arg1);
+          createNode(
```
Comment (nwangtw) [inline heron/statemgrs/src/java/com/twitter/heron/statemgr/zookeeper/curator/CuratorStateManager.java]: QQ: Do we need to delete the old node?

Comment (huijunw) [reply heron/statemgrs/src/java/com/twitter/heron/statemgr/zookeeper/curator/CuratorStateManager.java]: no, since the node is ephemeral

**Comment by huijunw** [comment]
@maosongfu updated with error/corner case handling

File: heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManager.java [hunk 2]
```diff
@@ -364,21 +369,26 @@ public void start() throws Exception {
...
-      metricsCacheManagerHttpServer.start();
-
-      // 2. The MetricsCacheServer would run in the main thread
-      // We do it in the final step since it would await the main thread
-      LOG.info("Starting Metrics Cache Server");
-      metricsCacheManagerServer.start();
-      metricsCacheManagerServerLoop.loop();
+      Boolean b = statemgr.setMetricsCacheLocation(metricsCacheLocation, topologyName)
+          .get(5000, TimeUnit.MILLISECONDS);
+      if (b != null && b == Boolean.TRUE) {
```
[the remark this answers is not in the document]
Comment (huijunw) [reply heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManager.java]: done

File: heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManager.java [hunk 3]
```diff
@@ -364,21 +369,26 @@ public void start() throws Exception {
...
+        LOG.info("Starting Metrics Cache HTTP Server");
+        metricsCacheManagerHttpServer.start();
+
+        // 2. The MetricsCacheServer would run in the main thread
+        // We do it in the final step since it would await the main thread
+        LOG.info("Starting Metrics Cache Server");
+        metricsCacheManagerServer.start();
+        metricsCacheManagerServerLoop.loop();
+      } else {
+        LOG.severe("statemgr metricscahe node failure, fail fast");
```
[the remark this answers is not in the document]
Comment (huijunw) [reply heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManager.java]: changed to exception

File: heron/statemgrs/src/java/com/twitter/heron/statemgr/zookeeper/curator/CuratorStateManager.java [hunk 4]
```diff
@@ -337,6 +340,26 @@ protected Lock getLock(String path) {
...
       TopologyMaster.MetricsCacheLocation location,
       String topologyName) {
+    client.getConnectionStateListenable().addListener(new ConnectionStateListener() {
+      @Override
+      public void stateChanged(CuratorFramework arg0, ConnectionState arg1) {
+        if (arg1 == ConnectionState.RECONNECTED || arg1 == ConnectionState.CONNECTED) {
+          LOG.info("zk session state changed to " + arg1);
+          ListenableFuture<Boolean> result = createNode(
+              StateLocation.METRICSCACHE_LOCATION, topologyName, location.toByteArray(), true);
+          try {
```
[the remark this answers is not in the document]
Comment (huijunw) [reply heron/statemgrs/src/java/com/twitter/heron/statemgr/zookeeper/curator/CuratorStateManager.java]: changed to fail fast when connection state is disrupted.

**Pull request merged**